src/magic_folder/test/test_snapshot.py

```diff
@@ -310,24 +378,117 @@ def test_serialize_deserialize_snapshot(self, content1, content2, filename):
         )
         d.addCallback(snapshots.append)
 
-        serialized = snapshots[1].to_json()
+        # serialize and store the snapshot in db.
+        # It should rewrite the previously written row.
+        self.db.store_local_snapshot(snapshots[1])
 
-        reconstructed_local_snapshot = LocalSnapshot.from_json(serialized, self.alice)
+        # now read back the serialized snapshot from db
+        reconstructed_local_snapshot = self.db.get_local_snapshot(filename, self.alice)
 
         self.assertThat(
             reconstructed_local_snapshot,
             MatchesStructure(
                 name=Equals(filename),
-                parents_local=HasLength(1),
+                parents_local=HasLength(1)
             )
         )
 
+        # the initial snapshot does not have parent snapshots
+        self.assertThat(
+            reconstructed_local_snapshot.parents_local[0],
+            MatchesStructure(
+                parents_local=HasLength(0),
+            )
+        )
+
+
+class TestRemoteSnapshot(AsyncTestCase):
+    """
+    Test upload and download of LocalSnapshot (creating RemoteSnapshot)
+    """
+
+    @inlineCallbacks
+    def setUp(self):
+        super(TestRemoteSnapshot, self).setUp()
+        self.root = create_fake_tahoe_root()
+        self.http_client = yield create_tahoe_treq_client(self.root)
+        self.tahoe_client = yield create_tahoe_client(
+            u"http://example.com",
+            self.http_client,
+        )
+        self.alice = create_local_author("alice")
+        self.stash_dir = mktemp()
+        os.mkdir(self.stash_dir)
+
+    def tearDown(self):
+        rmtree(self.stash_dir)
+        return super(TestRemoteSnapshot, self).tearDown()
+
+    def _download_content(self, snapshot_cap):
+        d = self.tahoe_client.download_capability(snapshot_cap)
+        data = json.loads(d.result)
+        content_cap = data["content"][1]["ro_uri"]
+        # sig = data["content"][1]["metadata"]["magic_folder"]["author_signature"]
+        # XXX is it "testtools-like" to check the signature here too?
+        return self.tahoe_client.download_capability(content_cap)
+
+    @given(
+        content=binary(min_size=1),
+        filename=magic_folder_filenames(),
+    )
+    def test_snapshot_roundtrip(self, content, filename):
+        """
+        Create a local snapshot, write into tahoe to create a remote snapshot,
+        then read back the data from the snapshot cap to recreate the remote
+        snapshot and check if it is the same as the previous one.
+        """
+        data = io.BytesIO(content)
+
+        snapshots = []
+        # create LocalSnapshot
+        d = create_snapshot(
+            name=filename,
+            author=self.alice,
+            data_producer=data,
+            snapshot_stash_dir=self.stash_dir,
```

Comment:
Why are we even using subclasses of test-case then? Can we just use pytest-style bare test-methods since we can't use any of the infrastructure of test-case anyway (we can't use setUp/tearDown and we aren't using their asserts .. and test-tools has plain function versions of everything right?)

In this particular case, it's probably a BETTER test if everything has to use the same stash-dir because many different things should be stashed there without overlapping / interfering, so Hypothesis using the same thing is a better test of reality I think.

Reply:
> Why are we even using subclasses of test-case then?

Do we need to have a broader discussion about overall testing strategy?

Reply:
> In this particular case, it's probably a BETTER test if everything has to use the same stash-dir because many different things should be stashed there without overlapping / interfering, so Hypothesis using the same thing is a better test of reality I think.

Chances are these won't be repeatable, though, because the implementation doesn't assign temporary names in a way Hypothesis controls - it uses `mkstemp` which is really just random, I think?  So we may get Hypothesis complaining about a "flaky" test because as far as it can tell the failure won't be reproducable. 

Also worth considering the performance characteristics of the test (sadly).  If we load up the stash directory with thousands of entries then filesystem performance may tank and the test may end up being very slow.

I think if we want to test handling of how collisions and such in the stash directory are handled it would be better to have explicit, dedicated tests for that purpose instead of getting them as a side-effect of these tests.  At the very least, if we get them as side-effects, we should have some assertions.  These tests won't necessarily fail if the implementation is bad because no assertions are made about state from previous iterations.  I think the implementation could write to the same stash file every time and this test would be satisfied because it only makes assertions about the current iteration's state.